codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/PaginatorGenerator.kt

```diff
@@ -211,9 +223,37 @@ class PaginatorGenerator private constructor(
                             "RuntimePlugins" to RuntimeType.runtimePlugins(runtimeConfig),
                         )
                     },
+                "is_empty_setter" to isEmptySetter(),
             )
         }
 
+    /** Generate code to calculate the value of is_empty. For most paginators this
+     * is indicated by the next token being the empty string. But for paginators
+     * with the isTruncatedPaginator trait the next token is not necessarily empty.
+     * (ex: for s3 ListParts the final next token is "0" when pagination is complete,
+     * causing the paginator to go back to the first page and loop forever)
+     * In this case we use a false value of isTruncated as the only indicator that
+     * the pagination is exhausted.
+     * */
+    private fun isEmptySetter() =
+        writable {
+            if (isTruncatedPaginator) {
+                rustTemplate(
+                    """
+                    // Pagination is exhausted when `is_truncated` is false
+                    let is_empty = !resp.is_truncated.unwrap_or(false);
```

Comment:
**nit**: This assumes the output shape member name is `is_truncated` which while likely true we could probably be more precise by having the trait include the member shape ID of the truncation member.